src/services/database_service.py

```
@@ -0,0 +1,535 @@
+            logger.warning(f"Batch delete failed: {e}. Trying one-by-one deletion as fallback.")
+            # Fallback: delete records one by one
+            for url in unique_urls:
+                try:
+                    self.supabase_client.table("crawled_pages").delete().eq("url", url).execute()
+                except Exception as inner_e:
+                    logger.warning(f"Error deleting record for URL {url}: {inner_e}")
+                    # Continue with the next URL even if one fails
+
+        # Check if contextual embeddings are enabled
+        use_contextual_embeddings = os.getenv("USE_CONTEXTUAL_EMBEDDINGS", "false") == "true"
+        logger.info(f"Use contextual embeddings: {use_contextual_embeddings}")
```

Pylint (3.3.7)

[warning] Use lazy % formatting in logging functions

(W1203)